curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1196,22 +1206,261 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
         if (ret == CURVEFS_ERROR::OK) {
             if (!iter->second->IsDirty()) {
                 AddReadDataCache(iter->second);
-                ReleaseWriteDataCache(iter->first, iter->second->GetLen());
+                ReleaseWriteDataCache(iter->first,
+                                      iter->second->GetActualLen());
             } else {
                 VLOG(6) << "data cache is dirty.";
             }
         } else if (ret == CURVEFS_ERROR::NOTEXIST) {
-            ReleaseWriteDataCache(iter->first, iter->second->GetLen());
+            ReleaseWriteDataCache(iter->first, iter->second->GetActualLen());
         }
     }
 
     return CURVEFS_ERROR::OK;
 }
 
-void DataCache::Write(uint64_t chunkPos, uint64_t len, const char *data,
-                      const std::vector<DataCachePtr> &mergeDataCacheVer) {
+DataCache::DataCache(S3ClientAdaptorImpl* s3ClientAdaptor,
+                     ChunkCacheManager* chunkCacheManager, uint64_t chunkPos,
+                     uint64_t len, const char* data)
+    : s3ClientAdaptor_(s3ClientAdaptor),
+      chunkCacheManager_(chunkCacheManager),
+      dirty_(true) {
+    uint64_t blockSize = s3ClientAdaptor->GetBlockSize();
+    uint32_t pageSize = s3ClientAdaptor->GetPageSize();
+    chunkPos_ = chunkPos;
+    len_ = len;
+    actualChunkPos_ = chunkPos - chunkPos % pageSize;
+
+    uint64_t headZeroLen = chunkPos - actualChunkPos_;
+    uint64_t blockIndex = chunkPos / blockSize;
+    uint64_t blockPos = chunkPos % blockSize;
+    uint64_t pageIndex, pagePos;
+    uint64_t n, m, blockLen;
+    uint64_t dataOffset = 0;
+    uint64_t tailZeroLen = 0;
+
+    while (len > 0) {
+        if (blockPos + len > blockSize) {
+            n = blockSize - blockPos;
+        } else {
+            n = len;
+        }
+        PageDataMap& pdMap = dataMap_[blockIndex];
+        blockLen = n;
+        pageIndex = blockPos / pageSize;
+        pagePos = blockPos % pageSize;
+        while (blockLen > 0) {
+            if (pagePos + blockLen > pageSize) {
+                m = pageSize - pagePos;
+            } else {
+                m = blockLen;
+            }
+
+            PageData* pageData = new PageData();
+            pageData->data = new char[pageSize];
+            if (!pageData->data) {
+                LOG(ERROR) << "new page failed.";
+                return;
+            }
```

Comment:
it doesn't work, when no memory is available,`new` will throw an exception and terminate the current process if we don't handle exceptions.
maybe, `new (std::nothrow) char[pageSize]` is you want, but if it doesn't throw here, it can throw anywhere when no memory available.